src/viz/layer/Layer.ts

```diff
@@ -510,14 +514,23 @@ export class Layer extends WithEvents implements StyledLayer {
       });
     }
 
-    let features = await this._viewportFeaturesGenerator.getFeatures();
-    const filters = this.filtersCollection.getApplicatorInstance(excludedFilters);
+    const isQueueEmpty = this._viewportFeaturesQueue.isQueueEmpty();
+    const pendingPromise = this._viewportFeaturesQueue.enqueue();
+
+    if (isQueueEmpty) {
+      this._viewportFeaturesGenerator.getFeatures().then(f => {
+        let features = f;
+        const filters = this.filtersCollection.getApplicatorInstance(excludedFilters);
 
-    if (this.filtersCollection.hasFilters()) {
-      features = features.filter(feature => filters.applicator(feature));
+        if (this.filtersCollection.hasFilters()) {
+          features = features.filter(feature => filters.applicator(feature));
+        }
+
+        this._viewportFeaturesQueue.resolveQueue(features);
+      });
     }
 
-    return features;
+    return (await pendingPromise.catch(error => console.debug(error))) || [];
```

Comment:
I'm wondering if we could hide a bit the complexity from the queue, under the ViewportFeaturesGenerator itself.... what do you think?

I'm thinking about instantiating _viewportFeaturesQueue inside it, and let the public method getFeatures make the IF-EMPTY-XXXX internally.... what do you think?

Reply:
I like this approach but we would need to refactor some things. I prefer to keep it as it is for the moment